README.rst

Django App Metrics
==================

.. image:: https://secure.travis-ci.org/frankwiles/django-app-metrics.png
    :alt: Build Status
    :target: http://travis-ci.org/frankwiles/django-app-metrics

django-app-metrics allows you to capture and report on various events in your
applications.  You simply define various named metrics and record when they
happen.  These might be certain events that may be immediatey useful, for
example 'New User Signups', 'Downloads', etc.

Or they might not prove useful until some point in the future.  But if you
begin recording them now you'll have great data later on if you do need it.

For example 'Total Items Sold' isn't an exciting number when you're just
launching when you only care about revenue, but being able to do a contest
for the 1 millionth sold item in the future you'll be glad you were tracking
it.

You then group these individual metrics into a MetricSet, where you define
how often you want an email report being sent, and to which User(s) it should
be sent.

Documentation
=============

Documentation can be found at ReadTheDocs_.

.. _ReadTheDocs: http://django-app-metrics.readthedocs.org/

Requirements
============

Celery_ and `django-celery`_ must be installed, however if you do not wish to
actually use Celery you can simply set ``CELERY_ALWAYS_EAGER = True`` in your
settings and it will behave as if Celery was not configured.

.. _Celery: http://celeryproject.org/
.. _`django-celery`: http://ask.github.com/django-celery/

Django 1.2 and above

Usage
=====

::

  from app_metrics.utils import create_metric, metric, timing, Timer, gauge

  # Create a new metric to track
  my_metric = create_metric(name='New User Metric', slug='new_user_signup')

  # Create a MetricSet which ties a metric to an email schedule and sets
  # who should receive it
  my_metric_set = create_metric_set(name='My Set',
                                    metrics=[my_metric],
                                    email_recipients=[user1, user2])

  # Increment the metric by one
  metric('new_user_signup')

  # Increment the metric by some other number
  metric('new_user_signup', 4)

  # Aggregate metric items into daily, weekly, monthly, and yearly totals
  # It's fairly smart about it, so you're safe to run this as often as you
  # like
  manage.py metrics_aggregate

  # Send email reports to users
  manage.py metrics_send_mail

  # Create a timer (only supported in statsd backend currently)
  with timing('mytimer'):
    for x in some_long_list:
       call_time_consuming_function(x)

  # Or if a context manager doesn't work for you you can use a Timer class
  t = Timer()
  t.start()
  something_that_takes_forever()
  t.stop()
  t.store('mytimer')

  # Gauges are current status type dials (think fuel gauge in a car)
  # These simply store and retrieve a value
  gauge('current_fuel', '30')
  guage('load_load', '3.14')

Backends
========

``app_metrics.backends.db`` (Default) - This backend stores all metrics and
aggregations in your database. NOTE: Every call to ``metric()`` generates a
database write, which may decrease your overall performance is you go nuts
with them or have a heavily traffic site.

``app_metrics.backends.mixpanel`` - This backend allows you to pipe all of
your calls to ``metric()`` to Mixpanel. See the `Mixpanel documentation`_
for more information on their API.

.. _`Mixpanel documentation`: http://mixpanel.com/docs/api-documentation

``app_metrics.backends.statsd`` - This backend allows you to pipe all of your
calls to ``metric()`` to a statsd server. See `statsd`_ for more information
on their API.

.. _`statsd`: https://github.com/etsy/statsd

``app_metrics.backends.redis`` - This backend allows you to use the metric() and
gauge() aspects, but not timer aspects of app_metrics.

``app_metrics.backends.librato_backend`` - This backend lets you send metrics to
Librato. See the `Librato documentation`_ for more information on their API.
This requires the `Librato library`_. It uses use a librato Gauge by default,
although this can be overridden by supplying ``metric_type="counter"`` as a
keyword arg to ``metric()``.

.. _`Librato documentation`: http://dev.librato.com/v1/metrics#metrics
.. _`Librato library`: http://pypi.python.org/pypi/librato/0.2

``app_metrics.backends.composite`` - This backend lets you compose multiple
backends to which metric-calls are handed. The backends to which the call is
sent can be configured with the ``APP_METRICS_COMPOSITE_BACKENDS`` setting. This
can be overridden in each call by supplying a ``backends`` keyword argument::

    metric('signups', 42, backends=['app_metrics.backends.librato',
                                    'app_metrics.backends.db'])


Settings
========

``APP_METRICS_BACKEND`` - Defaults to 'app_metrics.backends.db' if not defined.

``APP_METRICS_SEND_ZERO_ACTIVITY`` - Prevent e-mails being sent when there's been
no activity today (i.e. during testing). Defaults to `True`.

``APP_METRICS_DISABLED`` - If `True`, do not track metrics, useful for
debugging. Defaults to `False`.

Mixpanel Settings
-----------------
Set ``APP_METRICS_BACKEND`` == 'app_metrics.backends.mixpanel'.

``APP_METRICS_MIXPANEL_TOKEN`` - Your Mixpanel.com API token

``APP_METRICS_MIXPANEL_URL`` - Allow overriding of the API URL end point

Statsd Settings
---------------
Set ``APP_METRICS_BACKEND`` == 'app_metrics.backends.statsd'.

``APP_METRICS_STATSD_HOST`` - Hostname of statsd server, defaults to 'localhost'

``APP_METRICS_STATSD_PORT`` - statsd port, defaults to '8125'

``APP_METRICS_STATSD_SAMPLE_RATE`` - statsd sample rate, defaults to 1

Redis Settings
--------------
Set ``APP_METRICS_BACKEND`` == 'app_metrics.backends.redis'.

``APP_METRICS_REDIS_HOST`` - Hostname of redis server, defaults to 'localhost'

``APP_METRICS_REDIS_PORT`` - redis port, defaults to '6379'

``APP_METRICS_REDIS_DB`` - redis database number to use, defaults to 0

Librato Settings
----------------
Set ``APP_METRICS_BACKEND`` == 'app_metrics.backends.librato'.

``APP_METRICS_LIBRATO_USER`` - Librato username

``APP_METRICS_LIBRATO_TOKEN`` - Librato API token

``APP_METRICS_LIBRATO_SOURCE`` - Librato data source (e.g. 'staging', 'dev'...)

Composite Backend Settings
--------------------------
Set ``APP_METRICS_BACKEND`` == 'app_metrics.backends.composite'.

``APP_METRICS_COMPOSITE_BACKENDS`` - List of backends that are used by default,
e.g.::

    APP_METRICS_COMPOSITE_BACKENDS = ('librato', 'db', 'my_custom_backend',)

Running the tests
=================

To run the tests you'll need some requirements installed, so run::

    pip install -r requirements/test.txt

Then simply run::

    django-admin.py test --settings=app_metrics.tests.settings

TODO
----

- Improve text and HTML templates to display trending data well



docs/changelog.rst
=========
Changelog
=========

Version 0.8.0
-------------

    - Added Travis CI
    - Added librato backend
    - Added composite backends so you can send metrics to multiple backends automatically

Previous Versions
-----------------

Haven't really kept a strict history here, but we can recreate it from git logs.  In short
several contributors have added different backends such as statsd, redis, language translations,
and docs.

app_metrics/tasks.py
import base64
import json
import urllib
import urllib2
import datetime

try:
    from celery.task import task
except ImportError:
    from celery.decorators import task

from django.conf import settings
from django.core.exceptions import ImproperlyConfigured

from app_metrics.models import Metric, MetricItem, Gauge

# For statsd support
try:
    # Not required. If we do this once at the top of the module, we save
    # ourselves the pain of importing every time the task fires.
    import statsd
except ImportError:
    statsd = None

# For redis support
try:
    import redis
except:
    redis = None

# For librato support
try:
    import librato
    from librato.metrics import Gauge as LibratoGauge
    from librato.metrics import Counter as LibratoCounter
except ImportError:
    librato = None

class MixPanelTrackError(Exception):
    pass

# DB Tasks

@task
def db_metric_task(slug, num=1, **kwargs):
    met = Metric.objects.get(slug=slug)
    MetricItem.objects.create(metric=met, num=num)


@task
def db_gauge_task(slug, current_value, **kwargs):
    gauge, created = Gauge.objects.get_or_create(slug=slug, defaults={
        'name': slug,
        'current_value': current_value,
    })

    if not created:
        gauge.current_value = current_value
        gauge.save()


def _get_token():
    token = getattr(settings, 'APP_METRICS_MIXPANEL_TOKEN', None)

    if not token:
        raise ImproperlyConfigured("You must define APP_METRICS_MIXPANEL_TOKEN when using the mixpanel backend.")
    else:
        return token

# Mixpanel tasks

@task
def mixpanel_metric_task(slug, num, properties=None, **kwargs):

    token = _get_token()
    if properties is None:
        properties = dict()

    if "token" not in properties:
        properties["token"] = token

    url = getattr(settings, 'APP_METRICS_MIXPANEL_API_URL', "http://api.mixpanel.com/track/")

    params = {"event": slug, "properties": properties}
    b64_data = base64.b64encode(json.dumps(params))

    data = urllib.urlencode({"data": b64_data})
    req = urllib2.Request(url, data)
    for i in range(num):
        response = urllib2.urlopen(req)
        if response.read() == '0':
            raise MixPanelTrackError(u'MixPanel returned 0')


# Statsd tasks

def get_statsd_conn():
    if statsd is None:
        raise ImproperlyConfigured("You must install 'python-statsd' in order to use this backend.")

    conn = statsd.Connection(
        host=getattr(settings, 'APP_METRICS_STATSD_HOST', 'localhost'),
        port=int(getattr(settings, 'APP_METRICS_STATSD_PORT', 8125)),
        sample_rate=float(getattr(settings, 'APP_METRICS_STATSD_SAMPLE_RATE', 1)),
    )
    return conn


@task
def statsd_metric_task(slug, num=1, **kwargs):
    conn = get_statsd_conn()
    counter = statsd.Counter(slug, connection=conn)
    counter += num

@task
def statsd_timing_task(slug, seconds_taken=1.0, **kwargs):
    conn = get_statsd_conn()

    # You might be wondering "Why not use ``timer.start/.stop`` here?"
    # The problem is that this is a task, likely running out of process
    # & perhaps with network overhead. We'll measure the timing elsewhere,
    # in-process, to be as accurate as possible, then use the out-of-process
    # task for talking to the statsd backend.
    timer = statsd.Timer(slug, connection=conn)
    timer.send('total', seconds_taken)


@task
def statsd_gauge_task(slug, current_value, **kwargs):
    conn = get_statsd_conn()
    gauge = statsd.Gauge(slug, connection=conn)
    # We send nothing here, since we only have one name/slug to work with here.
    gauge.send('', current_value)

# Redis tasks

def get_redis_conn():
    if redis is None:
        raise ImproperlyConfigured("You must install 'redis' in order to use this backend.")
    conn = redis.StrictRedis(
        host=getattr(settings, 'APP_METRICS_REDIS_HOST', 'localhost'),
        port=getattr(settings, 'APP_METRICS_REDIS_PORT', 6379),
        db=getattr(settings, 'APP_METRICS_REDIS_DB', 0),
    )
    return conn

@task
def redis_metric_task(slug, num=1, **kwargs):
    # Record a metric in redis. We prefix our key here with 'm' for Metric
    # and build keys for each day, week, month, and year
    r = get_redis_conn()

    # Build keys
    now = datetime.datetime.now()
    day_key = "m:%s:%s" % (slug, now.strftime("%Y-%m-%d"))
    week_key = "m:%s:w:%s" % (slug, now.strftime("%U"))
    month_key = "m:%s:m:%s" % (slug, now.strftime("%Y-%m"))
    year_key = "m:%s:y:%s" % (slug, now.strftime("%Y"))

    # Increment keys
    r.incrby(day_key, num)
    r.incrby(week_key, num)
    r.incrby(month_key, num)
    r.incrby(year_key, num)

@task
def redis_gauge_task(slug, current_value, **kwargs):
    # We prefix our keys with a 'g' here for Gauge to avoid issues
    # of having a gauge and metric of the same name
    r = get_redis_conn()
    r.set("g:%s" % slug, current_value)

# Librato tasks

@task
def librato_metric_task(name, num, attributes=None, metric_type="gauge",
                        **kwargs):
    connection = librato.connect(settings.APP_METRICS_LIBRATO_USER,
                                 settings.APP_METRICS_LIBRATO_TOKEN)

    if metric_type == "counter":
        metric = LibratoCounter(connection, name, attributes=attributes)
    else:
        metric = LibratoGauge(connection, name, attributes=attributes)

    metric.add(num, source=settings.APP_METRICS_LIBRATO_SOURCE)
